plugins/upower.c

```diff
@@ -0,0 +1,276 @@
+			return FALSE;
+		}
+
+		dbus_message_iter_get_basic(&val, &percentage_val);
+		break;
+	}
+
+	if (percentage_val == 0.00) {
+		battery_val = 0;
+	} else if (percentage_val > 0.00 && percentage_val <= 100.00) {
+		battery_val = ((int) percentage_val - 1) / 20 + 1;
+        } else {
+		ofono_error("upower: Invalid value for battery level: %f",
+								percentage_val);
+		return FALSE;
+	}
+
+	DBG("done parsing upower message; last_battery_val: %d "
+							"battery_val: %d",
+							last_battery_val,
+							battery_val);
+	if (last_battery_val == battery_val)
+		return FALSE;
+	else
+		last_battery_val = battery_val;
+
+	for (list = modems; list; list = list->next) {
+		struct ofono_modem *modem = list->data;
+
+		__ofono_modem_foreach_registered_atom(modem,
+					OFONO_ATOM_TYPE_EMULATOR_HFP,
+					emulator_battery_cb,
+					GINT_TO_POINTER(battery_val));
+	}
+
+	return TRUE;
+}
+
+static void emulator_hfp_watch(struct ofono_atom *atom,
+				enum ofono_atom_watch_condition cond,
+				void *data)
+{
+	struct ofono_modem *modem = data;
+
+	if (cond == OFONO_ATOM_WATCH_CONDITION_REGISTERED) {
+		DBG("%s: REGISTERED", __func__);
+
+		modems = g_list_append(modems, modem);
+
+		if (modems->next != NULL)
+			return;
+
+		upower_battery_watch = g_dbus_add_signal_watch(connection,
+						UPOWER_SERVICE,
+						UPOWER_TOUCH_BATT_PATH,
+						DBUS_INTERFACE_PROPERTIES,
+						"PropertiesChanged",
+						battery_props_changed,
+						NULL, NULL);
+
+		/* register listener for battery events */
+	} else {
+		DBG("%s: UNREGISTERED", __func__);
+
+		modems = g_list_remove(modems, modem);
+		if (modems != NULL)
+			return;
+
+		g_dbus_remove_watch(connection, upower_battery_watch);
+	}
+}
+
+static void modemwatch(struct ofono_modem *modem, gboolean added, void *user)
+{
+
+	const char *path = ofono_modem_get_path(modem);
+
+	DBG("modem: %s, added: %d", path, added);
+
+	if (added) {
+		guint watch_id;
+
+		watch_id = __ofono_modem_add_atom_watch(modem,
+					OFONO_ATOM_TYPE_EMULATOR_HFP,
+					emulator_hfp_watch, modem, NULL);
+
+		g_hash_table_insert(modem_hfp_watches, g_strdup(path),
+						GINT_TO_POINTER(watch_id));
+
+	} else {
+		guint *watch_id = g_hash_table_lookup(modem_hfp_watches, path);
+
+		if (watch_id != NULL)
+			__ofono_modem_remove_atom_watch(modem, *watch_id);
+	}
+}
+
+static void call_modemwatch(struct ofono_modem *modem, void *user)
+{
+	modemwatch(modem, TRUE, user);
+}
+
+static void upower_connect(DBusConnection *conn, void *user_data)
+{
+	DBG("upower connect");
+
+	modem_watch = __ofono_modemwatch_add(modemwatch, NULL, NULL);
+	__ofono_modem_foreach(call_modemwatch, NULL);
+
+}
+
+static void upower_disconnect(DBusConnection *conn, void *user_data)
+{
+	DBG("upower disconnect");
+	__ofono_modemwatch_remove(modem_watch);
+}
+
+static int upower_init(void)
+{
+	DBG("upower init");
+
+	connection = ofono_dbus_get_connection();
+	upower_daemon_watch = g_dbus_add_service_watch(connection,
```

Comment:
Do we have a race condition here in case upower starts before ofono? Or upower_connect is called even in that case?